// @crl-rocks/crl-compiler/src/asts/instructions/If.ts
import {
  addError, compileTools,
  ControlStatement,
  executeStatements,
  Expression,
  RepresentTree,
  Statement,
} from "../AbstractTree";
import { CrlType } from "../../types/CrlType";
import { CrlBool } from "../../types/CrlBool";

export class If implements ControlStatement {
  readonly _condition: Expression;
  readonly _body: Statement[];
  readonly _elseBody?: Statement[];
  readonly rep: RepresentTree;

  _break?: boolean;
  _return?: CrlType;
  __return?: boolean;
  _continue?: boolean;

  constructor(
    condition: Expression,
    body: Statement[],
    elseBody?: Statement[]
  ) {
    this._condition = condition;
    this._body = body;
    this._elseBody = elseBody;

    this.rep = {
      type: "ControlStatement",
      represent: "Si",
      children: [
        { type: "Condition", represent: "Expresion" },
        {
          type: "Body",
          represent: "Cuerpo",
          children: body.map((instruct) => instruct.rep),
        },
      ],
    };
    if (elseBody) {
      this.rep.children?.push({
        type: "Statement",
        represent: "Sino",
        children: [
          {
            type: "Body",
            represent: "Cuerpo",
            children: elseBody.map((instruct) => instruct.rep),
          },
        ],
      });
    }
  }

  execute(): void {
    compileTools.scopeStack.push("SubAmbito_Si");
    this._condition.execute();

    if (this._condition._value) {
      try {
        const condition = this._condition._value.castTo(0) as CrlBool;
        if (condition.value) {
          executeStatements(this._body, this);
        } else if (this._elseBody) {
          compileTools.scopeStack.pop();
          compileTools.scopeStack.push("SubAmbito_Sino");
          executeStatements(this._elseBody, this);
        }
        compileTools.symbolsTable.removeScope(compileTools.scopeStack.length);
      } catch (e: any) {
        addError(this._condition, e.message);
      }
    }
    compileTools.scopeStack.pop();
  }
}
